server/src/main/java/com/cloud/resource/ResourceManagerImpl.java

```diff
@@ -1262,7 +1263,29 @@ private boolean doMaintain(final long hostId) {
                 return true;
             }
 
-            final List<HostVO> hosts = listAllUpAndEnabledHosts(Host.Type.Routing, host.getClusterId(), host.getPodId(), host.getDataCenterId());
+            List<HostVO> hosts = listAllUpAndEnabledHosts(Host.Type.Routing, host.getClusterId(), host.getPodId(), host.getDataCenterId());
+            if (hosts == null || hosts.isEmpty()) {
+                s_logger.warn("Unable to find a host for vm migration in cluster: " + host.getClusterId());
+                if (MIGRATE_VM_ACROSS_CLUSTERS.value()) {
+                    s_logger.info("Looking for hosts across different clusters in zone: " + host.getDataCenterId());
+                    hosts = listAllUpAndEnabledHosts(Host.Type.Routing, null, null, host.getDataCenterId());
+                    if (hosts == null || hosts.isEmpty()) {
+                        s_logger.warn("Unable to find a host for vm migration in zone: " + host.getDataCenterId());
+                        return false;
+                    }
+                    // Dont migrate vm if it has volumes on cluster-wide pool
+                    for (final VMInstanceVO vm : vms) {
+                        if (_vmMgr.checkIfVmHasClusterWideVolumes(vm.getId())) {
+                            s_logger.warn("Unable to migrate vm " + vm.getInstanceName() + " as it has volumes on cluster-wide pool");
+                            return false;
+                        }
+                    }
+                } else {
+                    s_logger.warn("Not migrating VM across cluster since " + MIGRATE_VM_ACROSS_CLUSTERS.key() + " is false");
+                    return false;
+                }
+            }
```

Comment:
sugestion:
`if (! clusterWideMigrationSupported(...)) return false;`
and extract this code to
`boolean clusterWideMigrationSupported(...)`